teps/0011-redirecting-step-inputs-and-outputs.md

```diff
@@ -0,0 +1,143 @@
+---
+title: redirecting-step-inputs-and-outputs
+authors:
+  - "@chhsia0"
+creation-date: 2020-08-17
+last-updated: 2020-08-17
+status: proposed
+---
+
+# TEP-0011: Redirecting Step Inputs and Outputs
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+    - [Example Usage](#example-usage)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+Consuming outputs of a step in another step is a common pattern in writing Tasks. However, this is currently tedious to do. Task authors have to overwrite the image entrypoint with either `sh -c` or the `script` field to wrap up the command to run with an explicit I/O redirection. This is not even possible if the image does not come with a shell, e.g., images built by [`ko`](https://github.com/google/ko).
+
+To achieve the functionality of I/O redirection between steps and even Tasks, Tekton will add new fields to steps for Task authors to specify paths to store stdout/stderr to and read stdin from.
+
+## Motivation
+
+Allow Task authors to support the following use cases:
+
+* Run image `gcr.io/k8s-staging-boskos/boskosctl` with args in a step and process its output through another `jq` image in a subsequent step to acquire the project name without overwriting the image entrypoint, building custom image with both utilities, or looking into container logs ([pipeline#2925](https://github.com/tektoncd/pipeline/issues/2925)).
+
+* Run image `alpine/git` with args `["describe", "--tags"]` in a step and extract the stdout as a Task result without overwriting the image entrypoint to be used by subsequent tasks.
+
+* In general, enable Task authors to access stdout/stderr of a step easily in another step in the same or different Task, without requiring authors to: 1) overwrite image entrypoints; 2) write wrapping scripts solely to do I/O redirection; and 3) build wrapper images to include all needed utilities.
+
+By enabling redirection between steps and Task results, it is much easier for Task authors to composite utility images in a Task.
+
+### Goals
+
+1. Allow the stdout/stderr of a step to be consumed by another step through stdin.
+1. Allow a step to produce a Task result through its stdout or stderr.
```

Comment:
This goal seems to add a lot of restrictions and new failure scenarios. Can we get by with just the first goal for the scope of this TEP?

Reply:
This goal seems important to me, especially if we're running an image without a shell and the command doesn't provide a flag to write output to a file. It also seems natural for user to find out if they can put stdout/stderr to results once we enable them to capture stdout/stderr.

On the other hand, users can always do:
```yaml
results:
- name command-output
steps:
- name: run-command-with-stdout
  image: ...
- name: copy-output-to-result
  image: alpine
  command: ["cp", "/tekton/steps/1/stdout", "$(results.command-output.path)"]
```
So we do have a workaround here.

Reply:
Rephrased the goal as suggesed by Scott.